# An introduction to data wrangling with Python and `pandas`

## Your name goes here. (Double-click on this cell to edit it.)

## Objectives

In this activity we will use the `pandas` library of Python tools to analyze a DataFrame. Learning outcomes include the following:

* You will become familiar with the Jupyter notebook interface and Markdown.
* You will import and use Python libraries.
* You will learn to load data into the Jupyter Notebook environment.
* You will practice some basic data cleaning and wrangling techniques.
* You will learn about the concept of "tidy" data.

### A note about the data in this activity

The data originally comes from [Gapminder](https://www.gapminder.org/) which describes it self as "a fact tank [that] fights devastating misconceptions about global development...making the world understandable based on reliable statistics". It has been processed lightly by the [Data Carpentry](http://www.datacarpentry.org/) community for data analysis activities. This activity is heavily based on the Data Exploration lesson developed by the Data Carpentry community.

## The Jupyter notebook

[Project Jupyter](http://jupyter.org/) is a non-profit, open-source project whose goal is to support interactive data science and scientific computing across all programming languages. According to their website,

> The Jupyter Notebook is an open-source web application that allows you to create and share documents that contain live code, equations, visualizations and narrative text. Uses include: data cleaning and transformation, numerical simulation, statistical modeling, data visualization, machine learning, and much more.

CoCalc supports Jupyter notebooks by allowing you to create and edit them in your browser while saving, tracking, and backing up your files in the cloud.

Jupyter notebooks are organized into cells. There are two types of cells we'll use:

1. Code cells:

In [1]:
# This is a code cell.
2+2

4

These are cells that run Python code. They are easy to spot because they have an input box labeled with `In []` with a number inside the brackets. The number represents the order in which the cells have been executed.

When you see the number sign `#` (or hashtag as the kids are calling it these days), that indicates a code "comment". This is text that appears inside a code cell, but will not be treated as code to run by Python. Of course that means that any code you *do* want to run needs to appear on its own line, without the number sign.

Click inside the cell above where it says `2+2`. Change it to `3+3`. Hit Ctrl-Enter. (Usually, the Mac equivalent of Ctrl-Enter is Cmd-Enter, but even on a Mac **you need to use Ctrl-Enter!**) This will execute the code in the cell again. Notice that when the code runs again, the number in brackets to the left updates. Don't worry too much about this number. The output (when there is output to be printed) will appear right below the input.

2. Markdown cells:

This is a Markdown cell. In fact, all the cells in this document that contain text are Markdown cells. To see the structure of this cell, double-click somewhere on top of this text. This will show you the plain text that generates the pretty, formatted text in the document. To get back to the formatted text, just hit Ctrl-Enter.

Double-clicking will also show you some formatting tricks that Markdown uses. After looking at the plain text, be sure to hit Ctrl-Enter to go back to the formatted text.

*italics*

**bold**

***bold italics***

~~strikethrough~~

`code`

> blockquotes

Horizontal rules:
***

[Web links](http://eelslap.com/)

* Bullet-
* pointed
* lists
  * with
  * indented
    * outline
    * capabilities

1. Numbered
2. Lists

Inline math: The quadratic fomula says $x = \frac{-b \pm \sqrt{b^{2} - 4ac}}{2a}$.

Displayed math:

$$i\hbar \frac{\partial}{\partial t}\lvert \Psi(\mathbf{r}, t) \rangle = \hat{H} \lvert \Psi(\mathbf{r}, t) \rangle$$

### Edit mode versus command mode: **VERY IMPORTANT!**

**It is very important that you double-click on Markdown cells before trying to edit them!** Double-clicking puts you in "edit" mode and allows you to type. You can tell you're in edit mode because the left edge of the cell is green. If you single-click only, the cell's left edge will be blue. This means you're still in "command" mode. In command mode, the keyboard is assigned to various tasks like creating or deleting cells. Doing this by accident is very bad because it messes up your document. **Again, do not type anything on the keyboard until you're absolutely sure you're in edit mode. (Look for the green edge on the left!)**

## Python

Python a popular programming language that is increasingly common in data science applications (and other sectors as well). The goal here is not to make you fluent in Python, or even really to teach you that much about computer programming. This assignment is to help familiarize you with some of the data processing tools found in Python.

You can learn more about Python here: https://www.python.org/

### Python libraries

A library in Python contains a set of tools (called functions) that perform tasks on our data. Instead of having to write all the functions yourself, someone else has written them for you.

Python doesn’t load all of the libraries available to it by default&mdash;that would be too inefficient. We have to add an `import` statement to our code in order to use library functions. To import a library, we use the syntax:

```
import libraryName
```

If we want to give the library a nickname to shorten the commands you need to type, we can do it like this:

```
import libraryName as nickname
```

Below is an example of importing the `pandas` library using the very common nickname `pd`. (Since we will be using `pandas`, this command occurs in a code cell later that we will actually run.)

```
import pandas as pd
```

### The `pandas` library

One of the best options for working with tabular data in Python is to use the Python Data Analysis Library (a.k.a. [pandas](http://pandas.pydata.org/)). The `pandas` library provides data structures, produces high quality plots, and integrates nicely with other libraries.

One of the best parts of the `pandas` library is that the developers have made it as easy to use as possible. The [documentation](http://pandas.pydata.org/pandas-docs/stable/index.html) is great, and they even made a [cheat sheet](https://github.com/pandas-dev/pandas/raw/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf) that has a lot of useful info!

#### Run the following code cell by clicking on it and then hitting Ctrl-Enter.

In [2]:
import pandas as pd

Each time we call a function that’s in a library, we use the following syntax:

```
LibraryName.FunctionName
```

Adding the library name with a `.` before the function name tells Python where to find the function. In the example above, we have imported `pandas` as `pd`. This means we don’t have to type out "`pandas`" each time we call a `pandas` function; we can just type `pd` as the `LibraryName`.

We will begin by locating and reading in our data in the section below.

## Loading data

### What is a DataFrame?

A DataFrame is a 2-dimensional data structure (rows and columns) that can store data of different types. It is similar to a spreadsheet or a data frame in R.

To use a `pandas` DataFrame, you'll need to read in the data from a file. The most common file format you'll encounter is a CSV ("comma separated values") file. Any program that allows you to manipulate data (like Excel, for example) will allow you to save/export files in the CSV format. From time to time, you'll have to read in files in a different format.

### The `gapminder` data

In your assignment folder is a file called `gapminderDataFiveYear_superDirty.txt`. This data is stored as a text file that happpens to be "tab separated" (meaning that the data values are separated by tabs rather than commas).

For CSV files, you use the `read_csv()` function from `pandas` to import data into Python. In our case, though, we'll use the more general `read_table()` function. We'll need an additional argument, `sep = '\t'` to tell the function to look for tabs between our data values.

#### Run the following code to read in the data and assign it to the variable `gapminder`.

In [3]:
# Run this code to assign the variable 'gapminder' to the entire dataset.
gapminder = pd.read_table("gapminderDataFiveYear_superDirty.txt", sep = '\t')

(The code worked silently in the background, so you should see no output. So how do you know if it worked? Check for the number on the left side of the cell in the brackets where it says `In []`. If there's a number there and you didn't otherwise get some kind of error message, then it worked.)

***

#### In the command above, why did we use `pd.read_table()`, not just `read_table()`? (Try `read_table` alone to see!)

**(Remember to double-click on the cell below&mdash;where it says "*Type your answer here*"&mdash;to put the worksheet into "edit" mode before you start typing your answer!)**

*Type your answer here.*

#### Could we use `pandas.read_table()` instead of `pd.read_table()`? (Again, try it and find out!)

*Type your answer here.*

***

Now we have a DataFrame called `gapminder` with all of our data in it! One of the first things we do after importing data is to have a quick look and make sure we see what we think we should see.

#### Type the name of the DataFrame in the cell below and run the code to see what happens.

In [4]:
# Type the name of your DataFrame on the next line and run this cell.
gapminder

year         pop  life Exp   gdpPercap            region
0     1952.0   8425333.0    28.801  779.445314  Asia_Afghanistan
1     1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
2     1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
3     1962.0  10267083.0    31.997  853.100710  Asia_Afghanistan
4     1962.0         NaN    31.997         NaN  Asia_Afghanistan
...      ...         ...       ...         ...               ...
1715  1987.0   9216418.0    62.351  706.157306   Africa_Zimbabwe
1716  1992.0  10704340.0    60.377  693.420786   Africa_Zimbabwe
1717  1997.0  11404948.0    46.809  792.449960   Africa_Zimbabwe
1718  2002.0  11926563.0    39.989  672.038623   Africa_Zimbabwe
1719  2007.0  12311143.0    43.487  469.709298   Africa_Zimbabwe

[1720 rows x 5 columns]

There are often too many rows to print to the screen and you don't really want to find out by printing them all to the screen and regretting it. By default, when you type the name of the DataFrame and run a cell, `pandas` knows not to print the whole thing if there are a lot of rows. Instead, you will see the first and last few rows with dots in between.

A neater way to look at a preview of the dataset is by using the `head()` method.  Writing `DataFrame.head()` will display the first five rows of the data. (This is another exmaple of the "dot" notation: since we want to apply the `head` function to our `DataFrame`, we use `DataFrame.head()`, replacing `DataFrame` with the name of our DataFrame.) You can specify how many rows you want to see as an argument, like `DataFrame.head(10)`. The `tail()` method does the same with the last rows of the DataFrame.

You do need to type the parenthesis at the end of the function, whether you put a number inside there or not.

#### Use these methods below to get an idea of what the `gapminder` DataFrame looks like.

In [5]:
# Add code here to view the head of the gapminder data.
gapminder.head(10)

year         pop  life Exp   gdpPercap            region
0  1952.0   8425333.0    28.801  779.445314  Asia_Afghanistan
1  1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
2  1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
3  1962.0  10267083.0    31.997  853.100710  Asia_Afghanistan
4  1962.0         NaN    31.997         NaN  Asia_Afghanistan
5  1967.0  11537966.0    34.020  836.197138  Asia_Afghanistan
6  1972.0  13079460.0    36.088  739.981106  Asia_Afghanistan
7  1977.0  14880372.0    38.438  786.113360  Asia_Afghanistan
8  1982.0  12881816.0    39.854  978.011439  Asia_Afghanistan
9  1987.0  13867957.0    40.822  852.395945  Asia_Afghanistan

In [6]:
# Add code here to view the tail of the gapminder data.
gapminder.tail(10)

year         pop  life Exp   gdpPercap           region
1710  1962.0   4277736.0    52.358  527.272182  Africa_Zimbabwe
1711  1967.0   4995432.0    53.995  569.795071  Africa_Zimbabwe
1712  1972.0   5861135.0    55.635  799.362176  Africa_Zimbabwe
1713  1977.0   6642107.0    57.674  685.587682  Africa_Zimbabwe
1714  1982.0   7636524.0    60.363  788.855041  Africa_Zimbabwe
1715  1987.0   9216418.0    62.351  706.157306  Africa_Zimbabwe
1716  1992.0  10704340.0    60.377  693.420786  Africa_Zimbabwe
1717  1997.0  11404948.0    46.809  792.449960  Africa_Zimbabwe
1718  2002.0  11926563.0    39.989  672.038623  Africa_Zimbabwe
1719  2007.0  12311143.0    43.487  469.709298  Africa_Zimbabwe

## Data wrangling

The term "data wrangling" describes the process of taking messy or disorganized data and cleaning it up. In this section (the bulk of this assignment), we'll learn about and execute commands to explore and describe our data with the aim of cleaning and organizing it.

### Counting rows and columns

We often want to know how many rows and columns are in the data&mdash;we want to know about the `shape` of the DataFrame. The `pandas` library has a convenient way for getting that information by using the following code:

```
DataFrame.shape
```

This returns a tuple (values separated by commas) representing the dimensions of the DataFrame (rows, columns).

Using `shape` does *not* require you to use parentheses at the end. (In fact, it won't work if you include parentheses.)

#### Write code using `shape` to get the shape of the `gapminder` DataFrame.

In [7]:
# Add code here to print the shape attribute of the gapminder data.
gapminder.shape

(1720, 5)

***

#### How many rows and columns are there?

*Type your answer here.*

***

The `info()` method gives a few useful pieces of information quickly, including the shape of the DataFrame, the variable type of each column, and the amount of memory stored.

(And here, we do need the parentheses again. It can be confusing when you're new to Python knowing when to use parentheses and when not to. Just follow the sample code provided and you'll be just fine for this assignment.)

#### Run the following code.

In [8]:
# Run this code to get info on the gapminder DataFrame.
gapminder.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1720 entries, 0 to 1719
Data columns (total 5 columns):
year         1717 non-null float64
pop          1715 non-null float64
life Exp     1719 non-null float64
gdpPercap    1718 non-null float64
region       1720 non-null object
dtypes: float64(4), object(1)
memory usage: 67.3+ KB


There are several problems with this data set as it currently stands. (This is normal; very few data sets are ready for data analysis right off the shelf.) The first step in data cleaning is identifying the problems. The `info()` above has already revealed one major problem.

Let's investigate!

***

#### Look at the number of entries in the entire DataFrame at the top of the `info()` output (`RangeIndex`) and compare that to each column. Explain what you think is happening. (Hint, go back and look at the spot earlier where you printed a good chunk of the data to the screen. Can you see spots where data might be "counted" or "uncounted" in the column totals?)

*Type your answer here.*

***

There are other fast, easy, and informative ways to get a sense of what your data might look like and any issues that it has. The `describe()` method will take the numeric columns and give a summary of their values. This is useful for getting a sense of the ranges of values and seeing if there are any unusual or suspicious numbers.

#### Use the `describe()` method on the gapminder DataFrame in the cell below.

In [9]:
# Add code here to describe each variable of the gapminder data.
gapminder.describe()

year           pop     life Exp      gdpPercap
count  1717.000000  1.715000e+03  1719.000000    1718.000000
mean   1979.449039  2.954973e+07    59.436493    7208.452200
std      17.228189  1.058366e+08    12.931041    9838.590984
min    1952.000000  6.001100e+04    23.599000     241.165877
25%    1967.000000  2.807799e+06    48.143000    1203.161887
50%    1977.000000  7.026113e+06    60.523000    3529.585686
75%    1992.000000  1.959366e+07    70.840500    9322.285886
max    2007.000000  1.318683e+09    82.603000  113523.132900

The DataFrame method `describe()` just blindly looks at all numeric variables. For example, we wouldn't actually want to find the mean value of `year`. Also, `describe()` ignores `NaN` ("Not a Number") values, which we might want to know about (and deal with, which we will later).

For now, let's look only at the columns that are truly continuous numbers (i.e., ignore `year`). For multiple columns, they need to be in double brackets as shown in the example below.

#### Run the following code.

In [10]:
# Run this code to describe numerical data (only columns with continuous values).
gapminder[["pop", "life Exp", "gdpPercap"]].describe()

pop     life Exp      gdpPercap
count  1.715000e+03  1719.000000    1718.000000
mean   2.954973e+07    59.436493    7208.452200
std    1.058366e+08    12.931041    9838.590984
min    6.001100e+04    23.599000     241.165877
25%    2.807799e+06    48.143000    1203.161887
50%    7.026113e+06    60.523000    3529.585686
75%    1.959366e+07    70.840500    9322.285886
max    1.318683e+09    82.603000  113523.132900

That's slightly better, but we still don't know about missing values yet.

The `value_counts()` method gives you a quick idea of what kinds of names are in your categorical variables. In this case our categorical data is the `region` column and represents the names of the regions where the data came from.

**Important observation: the data set covers 12 distinct years of data (every five years starting in 1952 and ending in 2007) for each `region`. Therefore, each `region` should appear 12 times.**

#### Use `value_counts()` on `gapminder` to see if all regions have 12 rows/entries as expected.

The command should be structured as follows, but replace `DataFrame` with the name of our data, and replace `column_name` with the name of the column we wish to analyze:

```
DataFrame["column_name"].value_counts()
```

#### Run the following code to check the number of observations for each region.

In [11]:
# Add code here using value_counts to check the number of observations for each region.
pd.value_counts(gapminder['region'])

Europe_Netherlands                         15
Americas_Dominican Republic                14
Asia_Afghanistan                           14
Americas_Bolivia                           14
Asia_Bangladesh                            13
                                           ..
    Asia_Korea, Dem. Rep.                   2
Africa_Cote d'Ivore                         1
    Asia_Jordan                             1
Africa_Congo, Democratic Republic           1
Africa_Democratic Republic of the Congo     1
Name: region, Length: 151, dtype: int64

**Uh-oh!** The table reveals several problems.

***

#### Describe and explain the (several) problems that you see.

*Type your answer here.*

***

### Handling missing data

Missing data is an important issue to handle. As we've seen, ignoring it doesn't make it go away, in our example, it has been giving us problems in our `describe()` results and our `value_counts()` results. (Remember the `NaN` values?)

There are different ways of dealing with missing data. There are pros and cons, so choose carefully:

* Analyze only the available data (i.e. ignore/delete the missing data).
* Replace the missing data with replacement values and treat these as though they were observed (*This is dangerous!*)
* Replace the missing data and account for the fact that these values were inputed with uncertainty. (For example, you could create a new true/false variable as a flag so you know that which values were not actually observed.)
* Use statistical models to allow for missing data that make assumptions about their relationships with the available data as necessary.

For our purposes with the dirty `gapminder` data set, it turns out that the rows with missing data are truly unnecessary, so we're going to choose to analyze only the available data. We will simply remove rows with missing values.

***

#### Why do we suspect the rows with missing data are truly unnecessary? Scan the printout of the `gapminder` data (earlier in the assignment) and explain why it appears that some regions have extra&mdash;and unnecessary&mdash;rows.

*Type your answer here.*

***

In `pandas` DataFrames, missing values are usually coded as `NaN` ("Not a Number") when a number is expected. A value like `NaN` is often called a "null" value.

Let's find out how many null values there are in the `gapminder` data. We are going to *chain* two steps together to determine the number of `NaN` values in the `gapminder` DataFrame.

* The `isnull()` method returns `True` or `False` for each cell in the DataFrame: `True` if the value in the cell is `NaN`, `False` if it is not.
* The `sum()` method adds up the `True` and `False` values in each column.

"Hold on," you say, "you can't add the words `True` and `False`!" Oh yes, you can! `True` is treated as 1 and `False` as 0 in Python (and in most programming languages), so if you `sum()` the results of `isnull()` you get the number of `NaN` values in each column. Awesome!

#### Run the chained code below.

In [12]:
# Run this code to count the number of null values.
gapminder.isnull().sum()

year         3
pop          5
life Exp     1
gdpPercap    2
region       0
dtype: int64

Yikes! There are null values in each column except region.

Removing null values from a DataFrame is incredibly easy to do because `pandas` allows you either to remove all instances with `NaN` or replace them with a particular value. (In fact, it may be too easy; make sure you are careful!)

The `dropna()` method drops rows with any column having null data. This method does not work "in place", meaning that it doesn't automatically save the changes to your existing DataFrame. So if you want to keep the modified DataFrame (without the null values), you need to assign it again like this:

```
DataFrame = DataFrame.dropna()
```

Or, you can preserve the original data by giving the result a new name:

```
DataFrame2 = DataFrame.dropna()
```

There are pros and cons to each approach. In this lab, we'll do both depending on the task.

You always want to be cautious about dropping rows from your data!

If you want to replace all null data with a `value` you can do this:

```
DataFrame = DataFrame.fillna(value)
```

Again, you have to assign the resulting output to a variable (either the same one or a different one) since `dropna()` and `fillna()` do not work "in place". If you don't assign the result to something, the output just prints to the screen, but nothing happens to the actual DataFrame stored in memory.

Before you rip into removing data, it's a good idea to know how many rows you have. You already know a few ways to do this.

#### Check the number of rows in the `gapminder` data.

In [13]:
# Add code here to check the number of rows in the gapminder data.
gapminder.shape[0]

1720

#### Now use `dropna()` to remove the null values from the `gapminder` DataFrame and assign the results to a new DataFrame called `gapminder2`.

In [14]:
## Add code here to use dropna() and assign the results to the DataFrame gapminder2.
gapminder2 = gapminder.dropna()

#### Now get the number of rows of `gapminder2`.

In [15]:
# Add code here to get the number of rows of gapminder2.
gapminder2.shape[0]

1712

#### Use `isnull()` and `sum()` (as  before) on `gapminder2` to confirm that there are no more null values.

In [16]:
# Add code here to check that there are no more null values.
gapminder2.isnull().sum()

year         0
pop          0
life Exp     0
gdpPercap    0
region       0
dtype: int64

***

#### How many rows were removed?

*Type your answer here.*

#### Which regions lost rows? Run some code to identify regions that still have too many observations. (Hint: look above to find the code you used to do this before, but remember that you are working with `gapminder2` now, not the original `gapminder`.)

In [17]:
# Add code here to print the number of observations per region.
gapminder2['region'].value_counts()

Americas_Dominican Republic                14
Europe_Netherlands                         14
Asia_Bangladesh                            13
Asia_Afghanistan                           13
Africa_Burundi                             13
                                           ..
Americas_Colombia                           2
    Asia_Jordan                             1
Africa_Congo, Democratic Republic           1
Africa_Cote d'Ivore                         1
Africa_Democratic Republic of the Congo     1
Name: region, Length: 151, dtype: int64

It looks like our work here is not yet done...

Before we go much further, we have to deal with the issue of the "index". When we printed the original `gapminder` DataFrame to the screen, the rows appeared with a number out to the left that started with 0 and ended at 1719 at the bottom. (Recall there were 1720 rows in the original data. And in Python&mdash;as in many programming languages&mdash;the counting starts with zero, not with one.) These numbers represent an *index* that keeps track of the row numbers. DataFrames always have an index that refers to the position of elements in the data structure. The index is useful for many of the `pandas` methods.

Let's look at the index of `gapminder2` now that we've removed the rows with null values.

#### Print the first ten rows of the `gapminder2` data.

In [18]:
# Add code here to print the first ten rows of gapminder2.
gapminder2.head(10)

year         pop  life Exp   gdpPercap            region
0   1952.0   8425333.0    28.801  779.445314  Asia_Afghanistan
1   1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
2   1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
3   1962.0  10267083.0    31.997  853.100710  Asia_Afghanistan
5   1967.0  11537966.0    34.020  836.197138  Asia_Afghanistan
6   1972.0  13079460.0    36.088  739.981106  Asia_Afghanistan
7   1977.0  14880372.0    38.438  786.113360  Asia_Afghanistan
8   1982.0  12881816.0    39.854  978.011439  Asia_Afghanistan
9   1987.0  13867957.0    40.822  852.395945  Asia_Afghanistan
10  1992.0  16317921.0    41.674  649.341395  Asia_Afghanistan

***

#### Examine the first ten rows of the index carefully. Notice anything unusual? Explain why the index appears the way it does.

*Type your answer here.*

***

In some situations, it might be helpful to know how the rows of `gapminder2` correspond to their original positions in the original `gapminder` DataFrame. However, in this case, we really only care about `gapminder2`. (The rows we removed did not represent real and accurate data.) So it will be more helpful to reset the index.

The `reset_index()` method will fix this. Note that `reset_index()`&mdash;like `dropna()`&mdash;requires us to reassign the results. Here, though, we will just overwrite `gapminder2` instead of creating a new variable. Also note the argument `drop = True` that tells Python not to keep the old index hanging around.

#### Run the following code to reset the index.

In [19]:
# Run this code to reset the index.
gapminder2 = gapminder2.reset_index(drop = True)

#### Print the first ten rows of `gapminder2` again to make sure this worked.

In [20]:
# Add code here to print the first ten rows of gapminder2.
gapminder2.head(10)

year         pop  life Exp   gdpPercap            region
0  1952.0   8425333.0    28.801  779.445314  Asia_Afghanistan
1  1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
2  1957.0   9240934.0    30.332  820.853030  Asia_Afghanistan
3  1962.0  10267083.0    31.997  853.100710  Asia_Afghanistan
4  1967.0  11537966.0    34.020  836.197138  Asia_Afghanistan
5  1972.0  13079460.0    36.088  739.981106  Asia_Afghanistan
6  1977.0  14880372.0    38.438  786.113360  Asia_Afghanistan
7  1982.0  12881816.0    39.854  978.011439  Asia_Afghanistan
8  1987.0  13867957.0    40.822  852.395945  Asia_Afghanistan
9  1992.0  16317921.0    41.674  649.341395  Asia_Afghanistan

### Remove duplicates

First, here's a nice trick: if you want to examine a *subset* of the rows&mdash;for example, to examine a region with too many observations&mdash;you can use code like this:

```
DataFrame[DataFrame.column == "value_in_column"]
```

where `column` is the name of a column and `value_in_column` is what you're interested in, for example the name of a specific region.

**Note the double equal sign:** single equal signs are used in Python to **assign** something to a variable (similar to `<-` in R if you are familiar with that). The double equal sign is a comparison: in this case, the value from the column on the left has to be *exactly* equal to the value string on the right.

#### Write code in the cell below to look at the region "Americas_Dominican Republic"

In [21]:
# Add code here to examine the rows from the "Americas_Dominican Republic" region.
gapminder2[gapminder2.region == 'Americas_Dominican Republic']

year        pop  life Exp    gdpPercap                       region
436  1952.0  2491346.0    45.928  1397.717137  Americas_Dominican Republic
437  1957.0  2923186.0    49.828  1544.402995  Americas_Dominican Republic
438  1962.0  3453434.0    53.459  1662.137359  Americas_Dominican Republic
439  1967.0  4049146.0    56.751  1653.723003  Americas_Dominican Republic
440  1967.0  4049146.0    56.751  1653.723003  Americas_Dominican Republic
441  1967.0  4049146.0    56.751  1653.723003  Americas_Dominican Republic
442  1972.0  4671329.0    59.631  2189.874499  Americas_Dominican Republic
443  1977.0  5302800.0    61.788  2681.988900  Americas_Dominican Republic
444  1982.0  5968349.0    63.727  2861.092386  Americas_Dominican Republic
445  1987.0  6655297.0    66.046  2899.842175  Americas_Dominican Republic
446  1992.0  7351181.0    68.457  3044.214214  Americas_Dominican Republic
447  1997.0  7992357.0    69.957  3614.101285  Americas_Dominican Republic
448  2002.0  8650322.0    70.847  4563.808154  Americas_Dominican Republic
449  2007.0  9319622.0    72.235  6025.374752  Americas_Dominican Republic

***

#### Look closely at the output above to determine why the "Americans_Dominican Republic" region has more than the 12 observations we are expecting?

*Type your answer here.*

***

Sometimes observations can end up in the data set more than once creating a duplicate row. Luckily, `pandas` has methods that allow us to identify which observations are duplicates. The method `df.duplicated()` will return `True` or `False` values for each row in the DataFrame telling you whether or not a row is an exact repeat.

#### Run the following code to see how it works.

In [22]:
## Run this code to get output for each row indicating if it's a duplicate row. (This will only look at the first five rows.)
gapminder2.duplicated().head()

0    False
1    False
2     True
3    False
4    False
dtype: bool

Since we don't want repeated rows (each country should only have one row per year), we can easily drop them with the `drop_duplicates()` method.

*Note:* The `drop_duplicates()` method is another one that does not work "in place". If you want to store the DataFrame without duplicate rows, you need to assign it again as follows:

```
DataFrame = DataFrame.drop_duplicates()
```

**Warning:** You should always be be cautious when dropping rows from your data.

***

#### Justify why it's okay for us to drop these rows.

*Type your answer here.*

***

#### Write code in the following cell to remove the duplicated rows from `gapminder2` and assign the resulting output back to `gapminder2`.

In [23]:
# Add code here to remove duplicate rows from gapminder2.
gapminder2 = gapminder2.drop_duplicates()

#### Use code to confirm that the first five rows no longer contain a duplicate row.

In [24]:
# Add code here to check the first five rows for duplicate rows.
gapminder2.duplicated().head(5)

0    False
1    False
3    False
4    False
5    False
dtype: bool

### Data types

Now that we have cleaned the data, we can take a close look at data types. Recall that the `info()` method shows us how the variables are stored:

#### Run the following code to check the data types for each variable.

In [25]:
# Run this code to check the data types for each variable.
gapminder2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1704 entries, 0 to 1711
Data columns (total 5 columns):
year         1704 non-null float64
pop          1704 non-null float64
life Exp     1704 non-null float64
gdpPercap    1704 non-null float64
region       1704 non-null object
dtypes: float64(4), object(1)
memory usage: 79.9+ KB


Notice that the `year` variable says `float64`. A "float" variable contains numbers that might include decimals. This is kind of meaningless for `year` since the year has to be a whole number. So let's change it to an "integer" data type. It's important to clean the data before trying to use the `astype()` method because `astype()` (along with many other `pandas` methods) will fail if there is missing data.

#### Run the following code to convert the `year` column to integers.

In [26]:
# Run this code to covert the year column to integers and check that it worked.
gapminder2["year"] = gapminder2["year"].astype("int")
gapminder2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1704 entries, 0 to 1711
Data columns (total 5 columns):
year         1704 non-null int64
pop          1704 non-null float64
life Exp     1704 non-null float64
gdpPercap    1704 non-null float64
region       1704 non-null object
dtypes: float64(3), int64(1), object(1)
memory usage: 79.9+ KB


***

#### Which other column should be stored as integers?

*Type your answer here*

***

#### Write code below to convert that column to an integer data type. In the same cell, use `info()` again to check that it worked.

In [27]:
# Add code here to convert another column to an integer data type and check that it worked.
gapminder2["pop"] = gapminder2["pop"].astype("int")
gapminder2.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1704 entries, 0 to 1711
Data columns (total 5 columns):
year         1704 non-null int64
pop          1704 non-null int64
life Exp     1704 non-null float64
gdpPercap    1704 non-null float64
region       1704 non-null object
dtypes: float64(2), int64(2), object(1)
memory usage: 79.9+ KB


### Handling Inconsistent Data

The `region` column still has issues that will affect our analysis.

#### Write code below to check the value counts again by region. (You've seen the command already above. Just copy it here.)

In [28]:
# Add code here to check the value counts by region.
gapminder2['region'].value_counts()

Asia_Vietnam                               12
Asia_Iraq                                  12
Africa_Morocco                             12
Asia_Afghanistan                           12
Africa_Uganda                              12
                                           ..
Americas_Colombia                           2
    Asia_Jordan                             1
Africa_Congo, Democratic Republic           1
Africa_Cote d'Ivore                         1
Africa_Democratic Republic of the Congo     1
Name: region, Length: 151, dtype: int64

***

#### What has been fixed about our value counts?

*Type your answer here.*

#### What still needs to be fixed about our value counts?

*Type your answer here.*

***

### String manipulations

Variables with text in them (called "character" or "string" variables) have some common problems:

* Lingering white space, usually at the end (which is not obvious when looking at the text).
* Uppercase and lowercase inconsistencies.

The latter issue is problematic because uppercase and lowercase characters are considered different characters by Python.

#### Run the following code to check this.

In [29]:
# Run this code to see that Python treats uppercase and lowercase differently.
"ABC" == "abc"

False

The following three methods are "string methods"(hence the `str`) that deal with these issues:

* `str.lstrip()` strips whitespace from the left.
* `str.rstrip()` strips whitespace from the right.
* `str.lower()` converts all characters to lowercase.

Here's how they look:

```
DataFrame["column_name"] = DataFrame["column_name"].str.lstrip()
DataFrame["column_name"] = DataFrame["column_name"].str.rstrip()
DataFrame["column_name"] = DataFrame["column_name"].str.lower()
```

*Note:* None of these methods work in place, so each needs to be stored back in the same column.

#### Write code that strips the white space from both sides of the values in the `region` column of `gapminder2` and makes all of the values lower case. Assign the results back to `gapminder2`.

In [30]:
# Add code here to strip white space on both left and right of region names and convert them to lowercase.
# (This will require three separate lines of code, but you can put all three lines in this cell.)

gapminder2['region'] = gapminder2['region'].str.strip().str.lower()
gapminder2['region'].value_counts()


europe_france                              12
africa_kenya                               12
asia_vietnam                               12
africa_djibouti                            12
africa_comoros                             12
                                           ..
americas_canada                             9
_canada                                     3
africa_congo, democratic republic           1
africa_cote d'ivore                         1
africa_democratic republic of the congo     1
Name: region, Length: 146, dtype: int64

#### Write code to check the `value_counts()` again to see the effect of the previous code cell.

In [31]:
# Add code here to check the counts for each region.
gapminder2['region'].value_counts()

europe_france                              12
africa_kenya                               12
asia_vietnam                               12
africa_djibouti                            12
africa_comoros                             12
                                           ..
americas_canada                             9
_canada                                     3
africa_congo, democratic republic           1
africa_cote d'ivore                         1
africa_democratic republic of the congo     1
Name: region, Length: 146, dtype: int64

***

#### Are we there yet? What is still wrong with the `region` column?

*Type your answer here.*

***

### Fixing the last few regions

At some point, data cleaning requires more manual intervention. Since we can see the regions that are recorded in error, we just need to fix them one at a time.

The `replace()` method is probably the easiest way to do that.

#### Run the following code to fix Canada.

In [32]:
# Run this code to fix Canada.
gapminder2["region"] = gapminder2["region"].replace("_canada", "americas_canada")

#### Now write several more lines of code to fix the remaining misrecorded regions.

In [33]:
# Add code here to fix the remaining misrecorded regions.


gapminder2['region'].replace("africa_cote.*", "africa_cote d'ivoire ", regex=True, inplace=True)

#### Print counts for each region one last time to make sure everything is truly fixed now.

In [34]:
# Add code here to print counts for each region.
gapminder2['region'].value_counts()

europe_france                              12
africa_kenya                               12
asia_vietnam                               12
africa_djibouti                            12
africa_comoros                             12
                                           ..
europe_iceland                             12
americas_jamaica                           12
africa_congo, dem. rep.                    10
africa_congo, democratic republic           1
africa_democratic republic of the congo     1
Name: region, Length: 144, dtype: int64

## Tidy data

Having what is called a *tidy* data set can make cleaning and analyzing your data much easier. Two of the important aspects of tidy data are:
* every variable has its own column,
* every observation has its own row.

There are other aspects of tidy data as well. Here is a good blog post about [tidy data in Python](http://www.jeannicholashould.com/tidy-data-in-python.html).

Currently the dataset has a single column for continent and country (the `region` column).


***

#### Why is having a single column for continent and country not tidy?

*Type your answer here*

***

### Let's make `gapminder` tidy

We can split the `region` column into two by using the underscore that separates continent from country.

We'll use the string method `split()` to split the string at the underscore (the first argument), which results in a list of two elements: before and after the underscore. The second argument, in this case `"1"`, tells `split()` that the split should take place only at the first occurrence of the underscore. Then the `str[]` specifies which item (either 0 or 1) to return.

So that we don't mess up the original `gapminder2` region column, let's create `gapminder3` as a home for our new columns.

#### Run the following code, which simply makes a copy of `gapminder2` and calls it `gapminder3`.

In [35]:
# Run this code to make a new copy of gapminder2 called gapminder3.
gapminder3 = gapminder2.copy()

#### Run the following code to split the current `region` column into proper `continent` and `country` columns.

In [36]:
# Run this code to split region into continent and country.
gapminder3["continent"] = gapminder3["region"].str.split('_', 1).str[0]
gapminder3["country"] = gapminder3["region"].str.split('_', 1).str[1]
gapminder3.head()

year       pop  life Exp   gdpPercap            region continent  \
0  1952   8425333    28.801  779.445314  asia_afghanistan      asia   
1  1957   9240934    30.332  820.853030  asia_afghanistan      asia   
3  1962  10267083    31.997  853.100710  asia_afghanistan      asia   
4  1967  11537966    34.020  836.197138  asia_afghanistan      asia   
5  1972  13079460    36.088  739.981106  asia_afghanistan      asia   

       country  
0  afghanistan  
1  afghanistan  
3  afghanistan  
4  afghanistan  
5  afghanistan

### Removing and renaming columns

We have now added the new columns `continent` and `country`, but we still have the old `region` column as well. In order to remove that column we use the `drop()` method. The first argument of `drop()` is the name of the variable to be dropped. The second argument is the *axis* number: 0 for row, 1 for column.

*Note:* any time we are getting rid of stuff, we want to make sure that we are doing it for a good reason and that we know our data will be okay after.

#### Run the following code to drop the superfluous `region` column.

In [37]:
# Run code here to drop the region column.
gapminder3 = gapminder3.drop('region', 1)
gapminder3.head()

year       pop  life Exp   gdpPercap continent      country
0  1952   8425333    28.801  779.445314      asia  afghanistan
1  1957   9240934    30.332  820.853030      asia  afghanistan
3  1962  10267083    31.997  853.100710      asia  afghanistan
4  1967  11537966    34.020  836.197138      asia  afghanistan
5  1972  13079460    36.088  739.981106      asia  afghanistan

Finally, it is a good idea to look critically at the column names themselves. It is a good idea to be as consistent as possible when naming columns. We often use all lowercase for all column names to avoid accidentally making names that can be confusing.

Avoiding spaces in column names will simplify your ability to manipulate data. Also look out for lingering white space at the beginning or end of your column names.

#### Run the following code that turns all column names to lowercase.

In [38]:
# Run this code to convert columns names to lowercase.
gapminder3.columns = gapminder3.columns.str.lower()
gapminder3.head()

year       pop  life exp   gdppercap continent      country
0  1952   8425333    28.801  779.445314      asia  afghanistan
1  1957   9240934    30.332  820.853030      asia  afghanistan
3  1962  10267083    31.997  853.100710      asia  afghanistan
4  1967  11537966    34.020  836.197138      asia  afghanistan
5  1972  13079460    36.088  739.981106      asia  afghanistan

We also want to remove the space from the `life exp` column name. We can do that with the `rename` method. It takes a *dictionary* as its argument, with the old column names as keys and new column names as values.

You're probably unfamiliar with dictionaries. They are a very useful data structure in Python, but we won't learn any more about them in this lesson. You can read more about them [here](https://docs.python.org/3/tutorial/datastructures.html#dictionaries).

#### Run the following code to rename the `life exp` column.

In [39]:
# Run this code to rename the life exp column.
gapminder3 = gapminder3.rename(columns={'life exp' : 'lifeexp'})
gapminder3.head()

year       pop  lifeexp   gdppercap continent      country
0  1952   8425333   28.801  779.445314      asia  afghanistan
1  1957   9240934   30.332  820.853030      asia  afghanistan
3  1962  10267083   31.997  853.100710      asia  afghanistan
4  1967  11537966   34.020  836.197138      asia  afghanistan
5  1972  13079460   36.088  739.981106      asia  afghanistan

***

#### Explain why the data set at this point is tidy, or at least much tidier than it was before.

*Type your answer here.*

***

### Export tidy data file

Now that we have a clean and tidy data set in a DataFrame, we want to export it and give it a new name so that we can refer to it and use it again later. The syntax is as follows:

```
DataFrame.to_csv("file_name_for_export.csv", index = False)
```

(The argument `index = False` is necessary unless you want the index to appear as its own column in your exported CSV file.)

#### Run the following code to export your clean and tidy DataFrame to a CSV file.

In [40]:
# Run this code to export the file gapminder_CandT.csv
gapminder3.to_csv("gapminder_CandT.csv", index = False)

## Conclusion

Now you know a little about the `pandas` library in Python. In particular, Python has a number of tools that make it relatively simple to clean and process data.

# Submit

Choose File -> Download as HTML 

Submit this HTML file to canvas.